rk4 integration for initial conditions

modichirag opened the pull request:
This pull request implements Runge Kutta integration for solving ODEs in growth functions when generating the cosmology object.

EiffL commented:
Thanks for opening this PR Chirag :-D

So first of all, yapf is the tool that enforces proper code formatting. From the root of your local clone run:
```
$ yapf -ir .
```
And if you don't already have yapf, you can just pip install it.

But then, looks like the tests are also failing

EiffL commented:
Yeah it's failing all over the place ^^ all the distance tests are failing

modichirag commented:
Curious. Because it turns out that I did not change the ODE in distance functions, I only changed it in the growth functions...

modichirag commented:
Very curious. I cloned the master branch anew and without changing anything, ran pytests on test_distancefunctions on my local machine. They fail, same as here. But reruning the tests on master here on github passes.

modichirag commented:
Hmm. So by mistake I had replaced the integration in distance ODE with DrummondPrince. Apparently that fails.
I have now replaced it with rk4 as well.
It still needs 256 points in integration for the raytracing checks to pass. But should be good now.

EiffL commented on flowpm/tfbackground.py:
```diff
@@ -334,12 +334,9 @@ def dchioverdlna(x, y, **kwcosmo):
-      dchioverdlna, tf.math.log(a)[0], 0.0, tf.math.log(a), constants=kwcosmo)
-  chitab = chitab.states[-1] - chitab.states
+  ## Run the ODE
+  chitab = rk4_integration(dchioverdlna, tf.math.log(a), 0.0, **kwcosmo)
```
we should also remove rtol from the set of arguments in this function

EiffL commented on flowpm/tfbackground.py:
```diff
@@ -496,6 +493,47 @@ def angular_diameter_distance(cosmo, a):
+  def body(i, yt, y):
+    t = a[i]
+    dt = da[i]
+    f1 = dydt(t, yt, **kwcosmo)
```
ah... well.... is that going to be properly diffable? @modichirag In many cases, any arguments you want to backprop over need to be tracked as inputs of the body function. Did you check if we were getting the correct gradients?

EiffL commented on flowpm/tfbackground.py:
```diff
@@ -539,64 +579,59 @@ def growth_ode(a, y, **kwcosmo):
 
 
 @tf.function
 def odesolve_func(a, rtol=1e-4, **kwcosmo):
```
we should remove rtol here as well

EiffL commented on flowpm/tfbackground.py:
```diff
@@ -609,6 +644,8 @@ def maybe_compute_ODE(cosmo, log10_amin=-2, steps=256):
     # been computed
     cache = cosmo._workspace['cache_ODE']
+    # Checking that the stored ODE results have the right lenght
+    assert len(cache['a']) == steps
```
hummm did you add that back? I think the assert was causing some issues inside tf.functions so w got rid of it

EiffL reviewed (commented):
This looks pretty nice, but I have some doubts the gradients are correct, see my comments. Could you check that we get the correct gradients with respect to cosmology?

modichirag commented:
I added a script verifying the analytic gradients with respect to \omega_c against numerical gradients. They match to within 0.1%. 
Also removed rtol factors and commented the assert out. I might have copied it from the previous version of the code by mistake.

dlanzieri reviewed (commented):
@EiffL is there an issue related to this pull request? I think I miss a bit of context

EiffL commented:
Yeah, it's this one: #82
